qiskit/quantum_info/operators/op_shape.py

```diff
@@ -0,0 +1,496 @@
+                self._num_qargs_l == other._num_qargs_l and
+                self._dims_r == other._dims_r and
+                self._dims_l == other._dims_l)
+
+    def copy(self):
+        """Make a deep copy of current operator."""
+        return copy.copy(self)
+
+    @property
+    def size(self):
+        """Return the combined dimensions of the object"""
+        return self._dim_l * self._dim_r
+
+    @property
+    def num_qubits(self):
+        """Return number of qubits if shape is N-qubit.
+
+        If Shape is not N-qubit return None
+        """
+        if self._dims_l or self._dims_r:
+            return None
+        if self._num_qargs_l:
+            if self._num_qargs_r and self._num_qargs_l != self._num_qargs_r:
+                return None
+            return self._num_qargs_l
+        return self._num_qargs_r
+
+    @property
+    def num_qargs(self):
+        """Return a tuple of the number of left and right wires"""
+        return self._num_qargs_l, self._num_qargs_r
+
+    @property
+    def shape(self):
+        """Return a tuple of the matrix shape"""
+        if not self._num_qargs_r:
+            # Vector shape
+            return (self._dim_l, )
+        # Matrix shape
+        return self._dim_l, self._dim_r
+
+    @property
+    def tensor_shape(self):
+        """Return a tuple of the tensor shape"""
+        return tuple(reversed(self.dims_l())) + tuple(reversed(self.dims_r()))
+
+    @property
+    def is_square(self):
+        """Return True if the left and right dimensions are equal."""
+        return (self._num_qargs_l == self._num_qargs_r
+                and self._dims_l == self._dims_r)
+
+    def dims_r(self, qargs=None):
+        """Return tuple of input dimension for specified subsystems."""
+        if self._dims_r:
+            if qargs:
+                return tuple(self._dims_r[i] for i in qargs)
+            return self._dims_r
+        num = self._num_qargs_r if qargs is None else len(qargs)
+        return num * (2, )
+
+    def dims_l(self, qargs=None):
+        """Return tuple of output dimension for specified subsystems."""
+        if self._dims_l:
+            if qargs:
+                return tuple(self._dims_l[i] for i in qargs)
+            return self._dims_l
+        num = self._num_qargs_l if qargs is None else len(qargs)
+        return num * (2, )
+
+    @property
+    def _dim_r(self):
+        """Return the total input dimension."""
+        if self._dims_r:
+            return reduce(mul, self._dims_r)
+        return 2 ** self._num_qargs_r
+
+    @property
+    def _dim_l(self):
+        """Return the total input dimension."""
+        if self._dims_l:
+            return reduce(mul, self._dims_l)
+        return 2 ** self._num_qargs_l
+
+    def validate_shape(self, shape):
+        """Raise an exception if shape is not valid for the OpShape"""
+        return self._validate(shape, raise_exception=True)
+
+    def valid_shape(self, shape):
```

Comment:
why are there 2 functions:` validate_shape` and `valid_shape` ? (it may be confusing for the user)
it seems that the only difference is whether it raises an exception or not.

Reply:
One returns True/False, the other raises an exception if its not valid. Currently only the exception one is used so `valid_shape` could be removed and added back later if it is ever needed